icu4c/source/common/putil.cpp

```diff
@@ -1171,16 +1171,16 @@ uprv_tzname(int n)
         because the tzfile contents is underspecified.
         This isn't guaranteed to work because it may not be a symlink.
         */
-        int32_t ret = (int32_t)readlink(TZDEFAULT, gTimeZoneBuffer, sizeof(gTimeZoneBuffer)-1);
-        if (0 < ret) {
+        char *ret = realpath(TZDEFAULT, gTimeZoneBuffer);
+        if (ret != nullptr && uprv_strcmp(TZDEFAULT, gTimeZoneBuffer) != 0) {
             int32_t tzZoneInfoTailLen = uprv_strlen(TZZONEINFOTAIL);
-            gTimeZoneBuffer[ret] = 0;
             char *  tzZoneInfoTailPtr = uprv_strstr(gTimeZoneBuffer, TZZONEINFOTAIL);
-
-            if (tzZoneInfoTailPtr != nullptr
-                && isValidOlsonID(tzZoneInfoTailPtr + tzZoneInfoTailLen))
-            {
-                return (gTimeZoneBufferPtr = tzZoneInfoTailPtr + tzZoneInfoTailLen);
+            if (tzZoneInfoTailPtr != nullptr) {
+                tzZoneInfoTailPtr += tzZoneInfoTailLen;
+                skipZoneIDPrefix(const_cast<const char **>(&tzZoneInfoTailPtr));
```

Comment:
Is this added line related to the same issue?
(I don't know my way around this part of the code...)

Reply:
Yes.
The path might be something like /usr/share/zoneinfo/posix/<OlsonID>.
The tzZoneInfoTailPtr will be put behind the "zoneinfo" part of the path. However, to fetch the OlsonID, we also need to get rid of the "posix" part. That's what skipZoneIDPrefix does. This function could be used without any changes.